src/pg_query_parse_plpgsql.c

```diff
@@ -354,13 +395,136 @@ PgQueryInternalPlpgsqlFuncAndError pg_query_raw_parse_plpgsql(CreateFunctionStmt
 	return result;
 }
 
+PgQueryInternalPlpgsqlFuncAndError pg_query_raw_parse_plpgsql_do(DoStmt* stmt)
+{
+	PgQueryInternalPlpgsqlFuncAndError result = {0};
+	MemoryContext cctx = CurrentMemoryContext;
+
+	char stderr_buffer[STDERR_BUFFER_LEN + 1] = {0};
+#ifndef DEBUG
+	int stderr_global;
+	int stderr_pipe[2];
+#endif
+
+#ifndef DEBUG
+	// Setup pipe for stderr redirection
+	if (pipe(stderr_pipe) != 0) {
+		PgQueryError* error = malloc(sizeof(PgQueryError));
+
+		error->message = strdup("Failed to open pipe, too many open file descriptors")
+
+		result.error = error;
+
+		return result;
+	}
+
+	fcntl(stderr_pipe[0], F_SETFL, fcntl(stderr_pipe[0], F_GETFL) | O_NONBLOCK);
+
+	// Redirect stderr to the pipe
+	stderr_global = dup(STDERR_FILENO);
+	dup2(stderr_pipe[1], STDERR_FILENO);
+	close(stderr_pipe[1]);
+#endif
+
+	PG_TRY();
+	{
+		result.func = compile_do_stmt(stmt);
+
+#ifndef DEBUG
+		// Save stderr for result
+		read(stderr_pipe[0], stderr_buffer, STDERR_BUFFER_LEN);
+#endif
+
+		if (strlen(stderr_buffer) > 0) {
+			PgQueryError* error = malloc(sizeof(PgQueryError));
+			error->message = strdup(stderr_buffer);
+			error->filename = "";
+			error->funcname = "";
+			error->context  = "";
+			result.error = error;
+		}
+	}
+	PG_CATCH();
+	{
+		ErrorData* error_data;
+		PgQueryError* error;
+
+		MemoryContextSwitchTo(cctx);
+		error_data = CopyErrorData();
+
+		// Note: This is intentionally malloc so exiting the memory context doesn't free this
+		error = malloc(sizeof(PgQueryError));
+		error->message   = strdup(error_data->message);
+		error->filename  = strdup(error_data->filename);
+		error->funcname  = strdup(error_data->funcname);
+		error->context   = strdup(error_data->context);
+		error->lineno    = error_data->lineno;
+		error->cursorpos = error_data->cursorpos;
+
+		result.error = error;
+		FlushErrorState();
+	}
+	PG_END_TRY();
+
+#ifndef DEBUG
+	// Restore stderr, close pipe
+	dup2(stderr_global, STDERR_FILENO);
+	close(stderr_pipe[0]);
+	close(stderr_global);
+#endif
+
+	return result;
+}
+
 typedef struct createFunctionStmts
 {
 	CreateFunctionStmt **stmts;
 	int stmts_buf_size;
 	int stmts_count;
 } createFunctionStmts;
 
+typedef struct doStmts
+{
+	DoStmt **stmts;
+	int stmts_buf_size;
+	int stmts_count;
+} doStmts;
```

Comment:
Thinking more about my other comment re: teaching `pg_query_raw_parse_plpgsql` to support both node types, I think you can actually greatly simplify this, by simply having one struct and one walker, that stores both CreateFunctionStmt and DoStmt nodes. Like this, renaming `createFunctionStmts` to `plStmts` for clarity:

```c
typedef struct plStmts
{
	Node **stmts;
	int stmts_buf_size;
	int stmts_count;
} plStmts;

static bool pl_stmts_walker(Node *node, plStmts *state)
{
	bool result;
	MemoryContext ccxt = CurrentMemoryContext;

	if (node == NULL) return false;

	if (IsA(node, CreateFunctionStmt) || IsA(node, DoStmt))
	{
		if (state->stmts_count >= state->stmts_buf_size)
		{
			state->stmts_buf_size *= 2;
			state->stmts = (Node**) repalloc(state->stmts, state->stmts_buf_size * sizeof(Node*));
		}
		state->stmts[state->stmts_count] = node;
		state->stmts_count++;
	} else if (IsA(node, RawStmt)) {
		return pl_stmts_walker((Node *) ((RawStmt *) node)->stmt, state);
	}

	PG_TRY();
	{
		result = raw_expression_tree_walker(node, pl_stmts_walker, (void*) state);
	}
	PG_CATCH();
	{
		MemoryContextSwitchTo(ccxt);
		FlushErrorState();
		result = false;
	}
	PG_END_TRY();

	return result;
}

PgQueryPlpgsqlParseResult pg_query_parse_plpgsql(const char* input)
{
	MemoryContext ctx = NULL;
	PgQueryPlpgsqlParseResult result = {0};
	PgQueryInternalParsetreeAndError parse_result;
	plStmts plStatements;
	size_t i;

	ctx = pg_query_enter_memory_context();

	parse_result = pg_query_raw_parse(input);
	result.error = parse_result.error;
	if (result.error != NULL) {
		pg_query_exit_memory_context(ctx);
		return result;
	}

	plStmts.stmts_buf_size = 100;
	plStmts.stmts = (Node**) palloc(plStmts.stmts_buf_size * sizeof(Node*));
	plStmts.stmts_count = 0;

	pl_stmts_walker((Node*) parse_result.tree, &plStmts);

	if (plStmts.stmts_count == 0) {
		result.plpgsql_funcs = strdup("[]");
		pg_query_exit_memory_context(ctx);
		return result;
	}

	result.plpgsql_funcs = strdup("[\n");

	for (i = 0; i < plStmts.stmts_count; i++) {
		PgQueryInternalPlpgsqlFuncAndError func_and_error;

		func_and_error = pg_query_raw_parse_plpgsql(plStmts.stmts[i]);
...
```

(and drop the whole portion where we deal with DO statements separately)